lib/ui/volatile_path_tracker.cc

```diff
@@ -11,75 +11,51 @@ VolatilePathTracker::VolatilePathTracker(
     bool enabled)
     : ui_task_runner_(ui_task_runner), enabled_(enabled) {}
 
-void VolatilePathTracker::Insert(std::shared_ptr<TrackedPath> path) {
+void VolatilePathTracker::Track(std::shared_ptr<TrackedPath> path) {
   FML_DCHECK(ui_task_runner_->RunsTasksOnCurrentThread());
   FML_DCHECK(path);
   FML_DCHECK(path->path.isVolatile());
   if (!enabled_) {
     path->path.setIsVolatile(false);
     return;
   }
-  paths_.insert(path);
-}
-
-void VolatilePathTracker::Erase(std::shared_ptr<TrackedPath> path) {
-  if (!enabled_) {
-    return;
-  }
-  FML_DCHECK(path);
-  if (ui_task_runner_->RunsTasksOnCurrentThread()) {
-    paths_.erase(path);
-    return;
-  }
-
-  std::scoped_lock lock(paths_to_remove_mutex_);
-  needs_drain_ = true;
-  paths_to_remove_.push_back(path);
+  paths_.push_back(path);
 }
 
 void VolatilePathTracker::OnFrame() {
   FML_DCHECK(ui_task_runner_->RunsTasksOnCurrentThread());
   if (!enabled_) {
     return;
   }
+#if !FLUTTER_RELEASE
   std::string total_count = std::to_string(paths_.size());
   TRACE_EVENT1("flutter", "VolatilePathTracker::OnFrame", "total_count",
                total_count.c_str());
-
-  Drain();
-
-  std::set<std::shared_ptr<TrackedPath>> surviving_paths_;
-  for (const std::shared_ptr<TrackedPath>& path : paths_) {
+#else
+  TRACE_EVENT0("flutter", "VolatilePathTracker::OnFrame");
+#endif
+
+  std::vector<std::weak_ptr<TrackedPath>> surviving_paths;
+  for (const std::weak_ptr<TrackedPath>& weak_path : paths_) {
```

Comment:
Instead of this pattern (which is totally fine IMO), an alternative is to use `std::remove_if`. I think the complexity is about the same (linear) but you don't have to allocate an intermediate. And, IMO, this way is a tad bit more cumbersome since you need to know the type of the container. `std::remove_if` only became available in C++17 so I don't think we've had the opportunity to use this in older code.

Reply:
Done